Feature/eric 43 update

Uli-Z opened the pull request:
# PR: Update to ERiC 43.3.2.0 and SDK Improvements

This PR updates the `eric-rs` library to support the latest **ERiC version 43.3.2.0** (Release 2025/2026) and introduces several improvements to the SDK's error handling and lifecycle management.

## Development Process & AI-Assisted Engineering

This update was developed using a structured AI-assisted engineering workflow to ensure high quality and compatibility despite breaking changes in the ERiC API:

1.  **Initial Code Review:** Performed by `gemini-3-flash-preview` to identify current limitations and map out the required changes for ERiC 43 compatibility.
2.  **Automated Binding Generation:** Used `bindgen` to generate new Rust bindings for the 43.3.2.0 header files.
3.  **Manual SDK Refactoring:** 
    *   Identification of breaking changes in versioned C-structures (removed fields `abrufCode` and `ersteSeite`).
    *   Fixing compiler errors caused by these API removals while maintaining safe memory alignment.
4.  **Feature Implementation:** Implementation of missing core features (`get_error_text`, `EricEntladePlugins`).
5.  **Extensive Testing:** 
    *   Validation of the library against the new 43.3.2.0 shared objects.
    *   Implementation of new integration tests (both positive and negative validation cases).
6.  **Final Quality Gate:** (Pending) Final architectural review by `gemini-3-pro-preview` to ensure idiomatic Rust patterns and robustness.

## Summary of Changes

### 1. Bindings Update
- Added pre-generated bindings for `ERiC 43.3.2.0 (Linux x86_64)`.
- Updated `eric-bindings/build.rs` to officially support and detect version 43.3.2.0.

### 2. SDK Fixes for ERiC 43
- **Breaking Changes in C-Structs:** Removed deprecated fields from parameter structures to match the ERiC 43 header definitions:
    - `eric_verschluesselungs_parameter_t`: Removed `abrufCode`.
    - `eric_druck_parameter_t`: Removed `ersteSeite`.

### 3. Feature Enhancements
- **Human-Readable Error Messages:** Implemented `get_error_text` (mapping to `EricHoleFehlerText`).
- **Integrated Error Reporting:** Operations now return German descriptive error texts directly from ERiC (e.g., "Fehler bei der Verarbeitung: ...").
- **Lifecycle Management:** Added `EricEntladePlugins` to `Drop` implementation for clean resource cleanup.
- **ErrorCode Completion:** Updated `ErrorCode` enum to include all 230 current error codes.

### 4. Testing
- Expanded integration tests in `test_validate.rs`.
- Added negative test cases for robust error handling verification.
- **Improved Test Infrastructure:** Transmission tests now gracefully skip with detailed setup instructions if test certificates or credentials are missing, preventing CI/local environment failures.
- Verified all tests pass using the ERiC 43.3.2.0 library.

## Verification
- `cargo test -p eric-sdk --test '*' -- --test-threads=1` passed successfully on Linux x86_64.

Uli-Z committed:
feat(bindings): add support for ERiC 43.3.2.0

Generierung der Rust-Bindings für die aktuelle ERiC-Version 43.3.2.0 (Linux x86_64).

Uli-Z committed:
fix(sdk): adapt structures to ERiC 43 changes

Entfernung der Felder 'abrufCode' und 'ersteSeite' aus den Parameter-Strukturen, da diese in ERiC 43 nicht mehr existieren.

Uli-Z committed:
feat(sdk): implement error text retrieval and plugin unloading

Hinzufügen von 'get_error_text' zur Ermittlung von Klartext-Fehlermeldungen und Aufruf von 'EricEntladePlugins' beim Beenden.

Uli-Z committed:
feat(sdk): update ErrorCode enum to match ERiC 43

Vervollständigung der Fehlercode-Enumeration auf insgesamt 230 Einträge gemäß der aktuellen Header-Datei.

Uli-Z committed:
test(sdk): add validation tests for ERiC 43

Erweiterung der Integrationstests um einen Fall für ungültiges XML zur Verifizierung der Fehlerberichterstattung.

Uli-Z committed:
test(sdk): add transmission tests and improve robustness

Einführung von Ende-zu-Ende Tests mit Test-Zertifikaten und automatisches Überspringen bei fehlender Umgebung.

quambene commented:
Thanks @Uli-Z for your contribution! First step to satisfy the CI pipeline is to update the env vars in `.github/workflows/ci.yml` to take the new version of the eric bindings into account

Uli-Z committed:
style: fix formatting and whitespace in eric-sdk

- added missing newlines at EOF
- formatted imports and long lines
- cleaned up trailing whitespace

Uli-Z committed:
fix(bindings): clean up build script and update docs-rs target

- removed unused variables in select_bindings()
- updated docs-rs binding selection to version 43.3.2.0

Uli-Z committed:
chore: ignore ERiC library directories

Added ERiC-*/ to .gitignore to prevent accidental commits of large
binary library files.

Uli-Z committed:
build: update versions and CI configuration to ERiC 43.3.2.0

- bumped eric-bindings to 0.5.0
- bumped eric-sdk to 0.4.0
- updated CI environment variables to point to the new ERiC version paths

Uli-Z commented:
Thanks for the quick response and your tip! I have updated the environment variables in .github/workflows/ci.yml to match the new ERiC version 43.3.2.0 as requested and bumped eric-bindings to 0.5.0 and eric-sdk to 0.4.0. (is this ok? maybe you want to change it...) I also performed local verification: All tests (including external-test mode) passed successfully. The CI checks should pass now (I hope so ;) )

quambene commented on eric-sdk/src/eric.rs:
```diff
@@ -215,13 +244,18 @@ impl Drop for Eric {
+                x if x == ErrorCode::ERIC_OK as i32 => (),
+                error_code => println!("Can't close eric: {}", error_code),
+            }
         }
```
minimize scope for `unsafe`:

``` suggestion
        let error_code = unsafe { EricEntladePlugins() };

        if error_code != ErrorCode::ERIC_OK as i32 {
            println!("Error while unloading plugins: {}", error_code);
        }

        let error_code = unsafe { EricBeende() };

        match error_code {
            x if x == ErrorCode::ERIC_OK as i32 => (),
            error_code => println!("Can't close eric: {}", error_code),
        }
```

quambene commented on eric-sdk/src/eric.rs:
```diff
@@ -201,6 +215,21 @@ impl Eric {
+        if error_code != ErrorCode::ERIC_OK as i32 {
+            let response_buffer = ResponseBuffer::new()?;
+            unsafe {
+                EricHoleFehlerText(error_code, response_buffer.as_ptr());
```
the todo above in line 207 can be removed now: `// TODO: EricHoleFehlerText() for error code`

quambene commented on eric-sdk/src/eric.rs:
```diff
@@ -201,6 +215,21 @@ impl Eric {
+                "Error during processing: {} ({})\nServer response: {}",
+                error_text,
+                error_code,
+                server_response
```
deserves a structured error at some point, but for the time being I'm suggesting to move the error code to the beginning of the error message, so that it's more visible:

``` suggestion
                "processing failed with error code {}: {}\nServer response: {}",
                error_code,
                error_text,
                server_response
    .with_context(|| format!("Server response:\n{}", server_response));
```

quambene commented on eric-sdk/tests/test_send.rs:
```diff
@@ -1,62 +1,102 @@
-fn test_send() {
-    let log_path = current_dir().unwrap();
-    let xml_path = Path::new("./test_data/taxonomy/v6.5/SteuerbilanzAutoverkaeufer_PersG.xml");
+fn setup_test_env() -> bool {
```
would prefer to avoid setting any environment variables, but instead make the env vars a requirement for the external tests. for example:

``` rust
fn require_test_env() -> Result<(), anyhow::Error> {
    env::var("CERTIFICATE_PATH").context("CERTIFICATE_PATH is required for external test")?;
    env::var("CERTIFICATE_PASSWORD").context("CERTIFICATE_PASSWORD is required for external test")?;
    Ok(())
}
```

further instructions can then be moved to the README in section "Test SDK".

quambene commented on eric-sdk/tests/test_send.rs:
```diff
@@ -1,62 +1,102 @@
         .context(format!("Can't read file: {}", xml_path.display()))
         .unwrap();
+
+    let hersteller_id = env::var("HERSTELLER_ID").unwrap_or_else(|_| "00000".to_string());
```
using an English term for "Hersteller ID", and make `VENDOR_ID` mandatory without fallback to make it more explicit:

``` suggestion
let vendor_id = env::var("VENDOR_ID").context("VENDOR_ID is required for external test")?;
```

quambene reviewed (commented):
looking good to me! made some minor suggestions for improvements

Uli-Z committed:
Apply review suggestions

Implemented feedback from PR #13:
- eric-sdk: Minimized unsafe blocks in Drop implementation
- eric-sdk: Moved error code to start of error message
- tests: Removed programmatic environment variable setup in test_send.rs
- docs: Added explicit instructions for running external tests with env vars

Uli-Z commented:
Thanks for the suggestions! I've implemented the changes:
- Minimized `unsafe` blocks in the `Drop` implementation for `Eric`.
- Updated the error message format to display the error code at the beginning.
- Refactored the integration tests to strictly require environment variables (`CERTIFICATE_PATH`, `CERTIFICATE_PASSWORD`, `VENDOR_ID`) instead of setting them programmatically.
- Updated the README with detailed instructions on how to provide these variables for external tests.
- Removed the obsolete TODO.

Everything is formatted and verified with `cargo check` and `clippy`.

quambene commented:
great work @Uli-Z ! I'm going to publish the new versions for `eric-bindings` and `eric-sdk` on [crates.io](https://crates.io/) in the upcoming days.
